File: ExerciseMicroservice/internal/auth/token_provider.go
package auth

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"
)

type ITokenProvider interface {
	IntrospectToken(bearerToken string) (bool, error)
	DecodeToken(token string) (map[string]interface{}, error)
	GenerateBasicAuthHeader() string
}

type TokenProvider struct {
	ClientID     string
	ClientSecret string
	Endpoint     string
}

func NewToken() ITokenProvider {
	clientID := getKeycloakClientId()
	clientSecret := getKeycloakClientSecret()
	endpoint := getKeycloakHost()

	tokenProvider := TokenProvider{
		ClientID:     clientID,
		ClientSecret: clientSecret,
		Endpoint:     endpoint,
	}

	return &tokenProvider
}

func (a *TokenProvider) IntrospectToken(bearerToken string) (bool, error) {
	authHeader := a.GenerateBasicAuthHeader()

	reqBody := fmt.Sprintf("token_type_hint=requesting_party_token&token=%s", bearerToken)
	req, err := http.NewRequest("POST", a.Endpoint, strings.NewReader(reqBody))
	if err != nil {
		return false, err
	}

	req.Header.Set("Authorization", authHeader)
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	client := http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return false, err
	}
	defer resp.Body.Close()

	// Check HTTP status code
	if resp.StatusCode != http.StatusOK {
		return false, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var introspectionResponse map[string]interface{}
	err = json.NewDecoder(resp.Body).Decode(&introspectionResponse)
	if err != nil {
		return false, err
	}

	active, ok := introspectionResponse["active"].(bool)
	if !ok {
		return false, fmt.Errorf("unexpected response format: %v", introspectionResponse)
	}

	return active, nil
}

func (a *TokenProvider) DecodeToken(token string) (map[string]interface{}, error) {
	// JWTs are typically in the format "header.payload.signature"
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil, fmt.Errorf("invalid token format")
	}

	// Decode the payload (second part)
	decodedPayload, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("error decoding token payload: %v", err)
	}

	// Unmarshal the JSON payload into a map
	var claims map[string]interface{}
	err = json.Unmarshal(decodedPayload, &claims)
	if err != nil {
		return nil, fmt.Errorf("error unmarshalling token claims: %v", err)
	}

	return claims, nil
}

func (a *TokenProvider) GenerateBasicAuthHeader() string {
	authString := fmt.Sprintf("%s:%s", a.ClientID, a.ClientSecret)
	authHeader := base64.StdEncoding.EncodeToString([]byte(authString))
	return fmt.Sprintf("Basic %s", authHeader)
}

func getKeycloakClientId() string {
	return os.Getenv("KEYCLOAK_CLIENT_ID")
}

func getKeycloakClientSecret() string {
	return os.Getenv("KEYCLOAK_CLIENT_SECRET")
}

func getKeycloakHost() string {
	return os.Getenv("KEYCLOAK_HOST") + "realms/" + os.Getenv("KEYCLOAK_REALM") + "/protocol/openid-connect/token/introspect"
}

File: OpenAIMicroservice/App/Services/Keycloak/token_provider.py
import os

import requests
import json
import base64

class TokenProvider:
    def __init__(self):
        self.endpoint = os.getenv("KEYCLOAK_HOST")
        self.client_id = os.getenv("KEYCLOAK_CLIENT_ID")
        self.client_secret = os.getenv("KEYCLOAK_CLIENT_SECRET")


    def introspect_token(self, bearer_token):
        auth_header = self.generate_basic_auth_header()

        req_body = f"token_type_hint=requesting_party_token&token={bearer_token}"
        headers = {
            "Authorization": auth_header,
            "Content-Type": "application/x-www-form-urlencoded"
        }

        try:
            response = requests.post(self.endpoint, data=req_body, headers=headers)
            response.raise_for_status()
        except requests.RequestException as e:
            return False, e

        try:
            introspection_response = response.json()
        except json.JSONDecodeError as e:
            return False, e

        active = introspection_response.get("active", False)
        return active, None

    def decode_token(self, token):
        parts = token.split(".")
        if len(parts) != 3:
            return None, ValueError("Invalid token format")

        try:
            decoded_payload = base64.urlsafe_b64decode(parts[1] + "==")
        except base64.binascii.Error as e:
            return None, e

        try:
            claims = json.loads(decoded_payload)
        except json.JSONDecodeError as e:
            return None, e

        return claims, None

    def generate_basic_auth_header(self):
        auth_string = f"{self.client_id}:{self.client_secret}"
        auth_header = base64.b64encode(auth_string.encode()).decode()
        return f"Basic {auth_header}"

